pkg/skaffold/config/portforward.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2019 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"fmt"
+	"strconv"
+)
+
+const (
+	compat   = "compat"
+	user     = "user"
+	services = "services"
+	debug    = "debug"
+	pods     = "pods"
+	off      = "off"
+)
+
+// PortForwardOptions are options set by the command line for port forwarding
+// with additional configuration information as well
+type PortForwardOptions struct {
+	Modes []string
+}
+
+// Enabled checks if the port-forwarding options indicates that forwarding should be enabled.
+func (p PortForwardOptions) Enabled() bool {
+	if len(p.Modes) == 0 {
+		return false
+	}
+	// --port-forward was previously a boolean, so accept pflag's boolean values.
+	// This method accepts "none" or bool variants to be mixed with others,
+	// leaving complaining to be done by Validate.
+	for _, o := range p.Modes {
+		b, err := strconv.ParseBool(o)
+		if err == nil && !b {
+			return false
+		}
+		if o == off {
+			return false
+		}
+	}
+	return true
+}
+
+// Validate checks that the port-forward options are ok.
+// For example, `none` is not mixed with other values.
+func (p PortForwardOptions) Validate() error {
+	// boolean values (true/false/1/0), `compat`, and `none` must be used alone
+	for _, o := range p.Modes {
+		if _, err := strconv.ParseBool(o); err == nil {
+			if len(p.Modes) > 1 {
+				return fmt.Errorf("port-forward %q cannot be combined with other options", o)
+			}
+		} else {
+			switch o {
+			case off, compat:
+				if len(p.Modes) > 1 {
+					return fmt.Errorf("port-forward %q cannot be combined with other options", o)
+				}
+			case user, services, pods, debug:
+				// continue
+			default:
+				return fmt.Errorf("unknown port-forward option %q: expected: user, services, pods, debug, none", o)
+			}
+		}
+	}
+	return nil
+}
+
+func (p PortForwardOptions) ForwardUser(runMode RunMode) bool {
+	for _, o := range p.Modes {
+		if b, err := strconv.ParseBool(o); err == nil && b {
+			o = compat
+		}
```

Comment:
This parsing is done multiple places. Can we add a`Parse` method which does this once and sets some values in `PortForwardOptions` struct.

Reply:
I had pursued this approach originally, to make `PortForwardOptions` implement pflag's `SliceValue` and not expose `PortForwardOptions.Modes`.  I hit some issues trying this and went for the easier route of keeping this `Modes`, but I'll take another stab.

Reply:
pflag's `Var` methods expect a `Value`; there's nothing that takes `SliceValue` like we'd need for the port-forward options.  I can make `PortForwardOptions` work for `Value`, but … it there are some other gotchas (the help text isn't updated).  So I'll work on this as a follow-on: there are some other bits that will require tweaking.